### PR TITLE
Fix/172 - Alter table to add constraints in a existent table - sqlite.

### PR DESCRIPTION
**Pull request by Marlysson**
Once that sqlite doesn't support define just a constraint in alter table statement there were some tests that set up just the foreign key constraint.

So, all times of alter table to add constraint ( of foreign key type ) should be set the new column that it are related to. I think it's. 

@josephmancuso please can you review this and if make sense?

### CODE REVIEW

**Comment by Marlysson** [comment]
**IN DRAFT**

**Comment by Marlysson** [comment]
@josephmancuso Done. 

Maybe need verify if wasn't setted foreign_keys to ON . But would be 2 queries just to do this. Then always I define it to True.
Tests: There's currently some way to do a integration tests?

Closes #172

File: src/masoniteorm/connections/SQLiteConnection.py [hunk 1]
```diff
@@ -34,10 +34,12 @@ def __init__(
 
     def make_connection(self):
         """This sets the connection on the connection class"""
-        print("db", self.database)
 
         self._connection = sqlite3.connect(self.database, isolation_level=None)
 
+        self._cursor = self._connection.cursor()
+        self._cursor.execute("PRAGMA foreign_keys = ON", ())
```
Comment (josephmancuso) [inline src/masoniteorm/connections/SQLiteConnection.py:41]: This should only be done when modifying the schema. We shouldn't turn pragma keys on every time we make a connection here

File: src/masoniteorm/connections/SQLiteConnection.py [hunk 2]
```diff
@@ -111,7 +113,7 @@ def query(self, query, bindings, results="*"):
             dict|None -- Returns a dictionary of results or None
         """
         query = query.replace("'?'", "?")
-        print("running query: ", self, query, bindings)
```
Comment (josephmancuso) [inline src/masoniteorm/connections/SQLiteConnection.py:114]: this is odd but this should stay for now. It helps me see what queries are running while testing

File: tests/sqlite/schema/test_sqlite_alter_table.py [hunk 3]
```diff
@@ -61,29 +65,53 @@ def test_can_alter_index(self):
...
+        )()
+        self.assertEqual(blueprint.to_sql(), sql)
+
     def test_can_alter_foreign_key(self):
         with self.schema.table("users") as blueprint:
+            blueprint.unsigned_integer("profile_id").nullable()
             blueprint.foreign("profile_id").references("id").on("profile")
 
-        sql = """ALTER TABLE "users" ADD CONSTRAINT users_profile_id_foreign FOREIGN KEY ("profile_id") REFERENCES "profile"("id")"""
+        sql = """ALTER TABLE "users" ADD "profile_id" INTEGER, CONSTRAINT users_profile_id_foreign REFERENCES "profile"("id")"""
```
Comment (josephmancuso) [inline tests/sqlite/schema/test_sqlite_alter_table.py:94]: I think these should have the pragma key string in them

**Review by josephmancuso — CHANGES_REQUESTED** [review]
I think the best way to go about this is to override the _compile_alter_table on the sqlite grammar class and do something like this:

```python
    def _compile_alter(self):
        """Compiles a query for altering table schemas.

        Returns:
            self
        """
        self._sql = "PRAGMA foreign_keys = ON; "
        self._sql += self.alter_format().format(
            table=self._compile_table(self.table),
            columns=self._compile_alter_columns(),
            constraints=self._compile_alter_constraints(),
            foreign_keys=self._compile_alter_foreign_keys(),
        )

        self._sql = self._sql.rstrip(", ")

        self._sql += "; PRAGMA foreign_keys = OFF; "



        return self
```

@girardinsamuel @Marlysson This way it only happens for sqlite grammars, only for alters and if we really need to only do this when adding constraints we can do an `if` check if the grammar has any constraints, then prepend and append the pragma thing

**Comment by josephmancuso** [comment]
I don't really understand this pragma thing actually. I'm reading it and it says each connection should turn pragma keys to on before executing a query in order to ensure foreign keys are honored? that sounds really dumb lol.

**Comment by josephmancuso** [comment]
I also don't think this actually works. When I run it in my client it just says query: OK but nothing happens

**Comment by josephmancuso** [comment]
I think the only way to do this is to actually run a bunch of queries. Like this:

```
https://stackoverflow.com/a/28869056
```

File: src/masoniteorm/connections/SQLiteConnection.py (hunk 1 above)
Comment (Marlysson) [reply src/masoniteorm/connections/SQLiteConnection.py:41]: You're right. I'll move to sqlite grammar.

**Comment by josephmancuso** [comment]
I think what we will have to do is fetch all the columns on the table, build a new grammar class out of them except this time add the new foreign key references and then drop the old table ....

**Comment by Marlysson** [comment]
> I don't really understand this pragma thing actually. I'm reading it and it says each connection should turn pragma keys to on before executing a query in order to ensure foreign keys are honored? that sounds really dumb lol.

Me neither. But seems that was problems with backwards compatibilities. Started with some behavior and its uses grow up so fast to possible changes.

> I also don't think this actually works. When I run it in my client it just says query: OK but nothing happens

I really need to do integration tests before push the changes..
But what operation do you tried?

**Comment by Marlysson** [comment]
> I think what we will have to do is fetch all the columns on the table, build a new grammar class out of them except this time add the new foreign key references and then drop the old table ....

Build a backup before add foreign keys or just copy table infos?

**Comment by Marlysson** [comment]
@josephmancuso based on what you said in twitter, this feature will by putted in stand by until conclude the stuffs in your tweet?

**Comment by josephmancuso** [comment]
@Marlysson I'm not sure. I think so? idt what's implemented in here actually works. At least it doesn't work for me. Whenever I try to run this:

```
ALTER TABLE "users" ADD "profile_id" INTEGER, CONSTRAINT users_profile_id_foreign REFERENCES "profile"("id")
```

It doesn't actually seem to work for me

**Comment by josephmancuso** [comment]
when I run it in table plus it doesn't seem to actually add the constraint

**Comment by Marlysson** [comment]
I've tried here in table plus too and it just create the relationship, but not the constraint.
You said that will take a look in changes in DLL SQL , are we try the snippet that you suggest in thread or wait to new solutions?

**Comment by josephmancuso** [comment]
@Marlysson take a look at some of these tests https://github.com/doctrine/dbal/blob/16878650d79934c31e27a13618ed6f3effee80d8/tests/Doctrine/Tests/DBAL/Platforms/SqlitePlatformTest.php#L297

**Comment by Marlysson** [comment]
@josephmancuso I'll put this PR in stand by because you are trying to redefine schema/grammar logic.

But I'm paying attention.

**Review by josephmancuso — APPROVED** [review]

**Comment by josephmancuso** [comment]
I love it but I will close in favor of #181

**Pull request closed without merging**